provider/balance_checker.go

```diff
@@ -2,153 +2,316 @@ package provider
+				// if there was provider restart with bunch of active leases
+				// spread their requests across 1min interval
+				// to reduce pressure on the RPC
+				if !ev.IsNewLease {
+					checkIn := time.Duration(rand.Int63n(int64(time.Minute))) // nolint: gosec
+					lState.tm = bc.timerFunc(ctx, checkIn, ev.LeaseID, false, leaseCheckCh)
+				} else {
+					bc.runEscrowCheck(ctx, ev.LeaseID, false, leaseCheckCh)
+				}
+			case event.LeaseRemoveFundsMonitor:
+				lsState, exists := bc.leases[ev.LeaseID]
+				if !exists {
+					break
+				}
+
+				if lsState.tm != nil && !lsState.tm.Stop() {
+					<-lsState.tm.C
+				}
+
+				delete(bc.leases, ev.LeaseID)
 			}
-		case withdrawAll := <-withdrawAllResult:
+		case res := <-leaseCheckCh:
+			// we may have timer fired just a heart beat ahead of lease remove event.
+			lState, exists := bc.leases[res.lid]
+			if !exists {
+				continue loop
+			}
+
+			withdraw := false
 
-			withdrawAllResult = nil
-			withdrawalTicker.Reset(bc.cfg.PollingPeriod) // Re-enable the timer
-			if err := withdrawAll.Error(); err != nil {
-				bc.log.Error("failed to started withdrawals", "err", err)
+			switch res.state {
+			case respStateScheduledWithdraw:
+				// reschedule periodic withdraw if configured
+				if bc.cfg.WithdrawalPeriod > 0 {
+					lState.scheduledWithdrawAt = time.Now().Add(bc.cfg.WithdrawalPeriod)
+				}
+				fallthrough
+			case respStateOutOfFunds:
+				withdraw = true
+				fallthrough
+			case respStateNextCheck:
+				timerPeriod := res.checkAfter
+				scheduledWithdraw := false
+				if res.err != nil {
+					bc.log.Info("couldn't check lease balance. retrying in 1m", "leaseId", res.lid, "error", res.err.Error())
+					timerPeriod = time.Minute
+				} else if !lState.scheduledWithdrawAt.IsZero() {
+					withdrawIn := lState.scheduledWithdrawAt.Sub(time.Now())
+					if timerPeriod >= withdrawIn {
+						timerPeriod = withdrawIn
+						scheduledWithdraw = true
+					}
+				}
+
+				lState.tm = bc.timerFunc(ctx, timerPeriod, res.lid, scheduledWithdraw, leaseCheckCh)
 			}
-		case <-withdrawalTicker.C:
-			withdrawAllNow = true
-			withdrawalTicker.Stop()
-		}
 
-		if withdrawAllNow {
-			bc.log.Info("balance below target amount, withdrawing now")
-			withdrawAllResult = runner.Do(func() runner.Result {
-				return runner.NewResult(nil, bc.startWithdrawAll())
-			})
+			if withdraw {
+				go func() {
+					select {
+					case <-bc.ctx.Done():
+					case resultCh <- runner.NewResult(res.lid, bc.startWithdraw(res.lid)):
+					}
+				}()
+			}
+		case res := <-resultCh:
+			if err := res.Error(); err != nil {
+				bc.log.Error("failed to do lease withdrawal", "err", err, "LeaseID", res.Value().(mtypes.LeaseID))
+			}
 		}
 	}
-	cancel()
 
 	bc.log.Debug("shutdown complete")
 }
+
+func (bc *balanceChecker) timerFunc(ctx context.Context, d time.Duration, lid mtypes.LeaseID, scheduledWithdraw bool, ch chan<- leaseCheckResponse) *time.Timer {
+	return time.AfterFunc(d, func() {
+		select {
```

Comment:
This is very similar to `bc.runEscrowCheck`. The primary difference is `case <- bc.lc.Done()` which is probably insignificant? Can this function be updated into something that just calls `bc.runEscrowCheck` ?